stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb

```diff
@@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
     return base._copyToNativeArrayBuffer()
   }
 
-  /// Copy a Sequence into an array.
-  public func _initializeTo(ptr: UnsafeMutablePointer<Base.Generator.Element>)
-    -> UnsafeMutablePointer<Base.Generator.Element> {
+  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
+  /// returning one past the last element initialized, or a `nil` pointer
+  /// if the copy cannot be safely made given `capacity`.
+  public func _initializeTo(
```

Comment:
`capacity` is the number of uninitialized elements that the caller made available in the buffer that `ptr` points to; indeed, this should be spelled out in the doc comment. 

The parameter is necessary because `_initializeTo` is defined on `SequenceType`, where there is no guaranteed `count` — so the caller might not be able to allocate enough space to fit all elements. An alternative would be to trust that the caller allocates enough space for `underestimatedCount` elements, but that didn't feel like the right API to me. Having the capacity explicitly supplied is harder to get wrong, and (in theory) allows the function to work in cases when the underestimate was low but the caller happened to allocate more space anyway due to e.g. fitting storage size on an exponential growth curve.

My initial instinct was to move the definition down to `CollectionType`, but @gribozavr didn't like that:

>> Would you be against a patch that removed _initializeTo from SequenceType and moved it into CollectionType instead?

> I think that's a reasonable start, but not a perfect long-term solution.  The issue is that, Array() initializer takes a sequence (as do other functions that would use initalizeTo(), like appendContentsOf()).  We can add an overload for collections, but it is not a great solution.  Sometimes happens that a collection is statically known to be only a sequence (for example, when the Array() initializer is called from a generic function).  That would force collections to go through a slower code path.

> Another reasonable thing to do would be to allow _initializeTo to somehow refuse to operate, if it is not dynamically a collection, and then have a fallback on the caller side.  See _customContainsEquatableElement and how it is used for an example of this trick.

I think the argument for allowing `_initializeTo` to work in `SequenceType` context is a good one. I went with Dmitri's suggestion and allowed the method to fail; I added the capacity parameter during the course of the implementation.

* * *

The `_initializeTo` in this PR is somewhat limited by only allowing the sequence to be copied into a single contiguous buffer. This pretty much limits its use to `Array`. I'd actually prefer to see a non-failing but continuable variant that would allow its use in any collection that segments its storage into multiple contiguous regions (e.g., a deque based on a circular buffer, or even my B-tree collections). This would make the method signature even more complicated in `Sequence`, though. OTOH, it would fit a bit better into `IteratorProtocol`, with a declaration that could be a variant of this:

```
/// Copy at most `capacity` elements starting from the current element into
/// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
/// - Returns: `true` if there are more elements remaining in the sequence.
mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
```

(I don't like the `inout` parameter or the boolean result though.) However, having `_copyContents` in `IteratorProtocol` would probably mean we'd need to change `Array` to use a specialized iterator other than `IndexingIterator`. This would mean writing up a public API proposal that'd be hard to justify unless `copyContents` was made public at the same time — and I'm not yet sure if the speedup would be significant enough to be worth that effort.

I'd  be OK with any of the following options:

1. Leave the `capacity` parameter in place (possibly with a better label), but explicitly document what it does.
2. Drop the parameter, but document that the caller is supposed to supply enough space to fit `underestimatedCount` elements. The method would still need to be able to fail, though.
3. Move the method down to `Collection`, restoring it to the original signature, but making it less useful.
4. Drop the `_initializeTo` part of this PR, and concentrate on replacing indexing with iterators (along with speeding up iterators when necessary).
5. Do more work to convert `_initializeTo` into a restartable/continuable thing, possibly moving it into `IteratorProtocol`.

Thoughts?